src/jit/gentree.h

```diff
@@ -1747,6 +1753,15 @@ struct GenTree
     // sets "*pIsEntire" to true if this assignment writes the full width of the local.
     bool DefinesLocalAddr(Compiler* comp, unsigned width, GenTreeLclVarCommon** pLclVarTree, bool* pIsEntire);
 
+    regNumber RegAtDef() const
+    {
+        return GetRawRegNum();
+    }
+    regNumber RegAtUse() const
+    {
+        return isContained() ? REG_NA : GetRawRegNum();
+    }
+
```

Comment:
I personally find gtRegNum a simple abstraction to get the reg allocated to a gentree node (whether it is def or use).
Prior to regOptional feature gtRegNum consistently can be used to know the reg allocated to a contained or non-contained node.  Because for contained node, it is set to REG_NA and non-contained node it will have a valid regNum.  This fact codegen used to directly use gtRegNum irrespective of whether an operand is contained or not.

But regOptional spillTemps (untracked locals and Gentree nodes with no backing stack locations like GT_ADD) pose an issue that gtRegNum will represent the reg at Def but not at Use if it becomes a contained spill temp.   For such operands, if codegen depended on gtRegNum will be REG_NA, it will lead to SBCG.  We don't have this issue for contained memory operand cases since gtRegNum on those nodes will be set to REG_NA both at Def and at Use.

In summary codegen needs to guard against using gtRegNum in case of contained spill temps at their use.  It is easy to detect by adding the following code in GetRegNum

if (isContainedSpillTemp())
{
   return REG_NA
}
else
{
  // existing logic
}

Node that if LSRA decides not to give a reg to a reg-optional tree temp at its use, LSRA will set GT_NOREG_AT_USE flag on tree node.  But such a tree node will not be considered contained spill temp, till it gets spilled at its def (i.e. GTF_SPILLED flag is set).  In essence the above if-check will not consider a spill temp as contained at its def when querying gtRegNum.  After the codegen of tree temp at its def and it was spilled, later when gtRegNum is queried on, since GTF_SPILLED nad GT_NOREG_AT_USE flags are set it will be considered a spill temp and will return REG_NA.